### ISSUE TITLE
torch.topk returned values and indices are reordered if sorted=False

### ISSUE THREAD

**Issue by PassengerC07** [pytorch/pytorch#108642]
### 🐛 Describe the bug

If sorted=False, torch.topk will return a set of reordered values and indices tensors instead of the original values

![topk](https://github.com/pytorch/pytorch/assets/20369971/55631820-62d2-4eb4-9544-38268554e77c)

### Versions

 
Collecting environment information...
PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Rocky Linux release 8.5 (Green Obsidian) (x86_64)
GCC version: (GCC) 10.2.0
Clang version: Could not collect
CMake version: version 3.27.2
Libc version: glibc-2.28

Python version: 3.8.0 | packaged by conda-forge | (default, Nov 22 2019, 19:11:38)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-348.el8.0.2.x86_64-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A40
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              192
On-line CPU(s) list: 0-191
Thread(s) per core:  2
Core(s) per socket:  48
Socket(s):           2
NUMA node(s):        2
Vendor ID:           AuthenticAMD
CPU family:          23
Model:               49
Model name:          AMD EPYC 7552 48-Core Processor
Stepping:            0
CPU MHz:             2200.000
CPU max MHz:         2200.0000
CPU min MHz:         1500.0000
BogoMIPS:            4400.06
Virtualization:      AMD-V
L1d cache:           32K
L1i cache:           32K
L2 cache:            512K
L3 cache:            16384K
NUMA node0 CPU(s):   0-47,96-143
NUMA node1 CPU(s):   48-95,144-191
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] numpy==1.24.4
[pip3] torch==2.0.1
[pip3] torch-geometric==2.3.1
[pip3] torch-scatter==2.1.1
[pip3] triton==2.0.0
[conda] numpy                     1.24.4                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torch-geometric           2.3.1                    pypi_0    pypi
[conda] torch-scatter             2.1.1                    pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by cpuhrsch** [comment]
Is your expected behavior that topk becomes an identity function if k equals n (the length of the input)?

Do you have some more detail on the usecase here?

**Comment by PassengerC07** [comment]
> Is your expected behavior that topk becomes an identity function if k equals n (the length of the input)?
> 
> Do you have some more detail on the usecase here?

I'm trying to use topk to extract individual rankings of two tensors, with which I can compute the target in torch.nn.MarginRankingLoss(input1, input2, target). The rankings in the inputs in my case are based on the magnitude of individual elements' values. The behavior of topk with sorted=False makes this difficult.

I can use only the "TopK" portion of topk output to compute this and miss the last bit of the tensor, as it seems that only the last element is miss placed. If I set k=tensor.shape[0]-1, the returned values and indices are correct.

:D Omitting one case when computing loss should be okay but my OCD is screaming.

**Comment by PassengerC07** [comment]
Come to think about it, my needs can be satisfied by using torch.argsort. But still this miss-placed last element in torch.topk is a bit troublesome.

**Comment by cpuhrsch** [comment]
> But still this miss-placed last element in torch.topk is a bit troublesome.

This isn't immediately obvious to me from your example. What exactly is misplaced here and what is the expected behavior?
